File: android/app/src/main/java/tw/speedalert/data/EnforcementStore.kt
package tw.speedalert.data

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import tw.speedalert.model.EnforcementItem
import tw.speedalert.model.EnforcementKind
import tw.speedalert.model.NearbyHit
import tw.speedalert.util.Geo
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.max

/**
 * Loads the bundled offline dataset (`assets/app_data.json`) and answers
 * proximity queries via a ~2.2 km grid index. Mirrors iOS `EnforcementStore`.
 */
class EnforcementStore(private val context: Context) {

    var items: List<EnforcementItem> = emptyList(); private set
    var builtAt: String? = null; private set
    var loadError: String? = null; private set

    private val cell = 0.02
    private val metersPerDegLat = 111_320.0
    private val grid = HashMap<Long, MutableList<Int>>()

    init { load() }

    private fun load() {
        try {
            val text = context.assets.open("app_data.json").bufferedReader().use { it.readText() }
            val root = JSONObject(text)
            builtAt = if (root.isNull("built_at")) null else root.optString("built_at")

            val list = ArrayList<EnforcementItem>()
            addCameras(root.optJSONArray("cameras"), list)
            addSections(root.optJSONArray("sections"), list)
            addAccidents(root.optJSONArray("accidents"), list)

            items = list
            buildIndex()
            loadError = null
            Log.i(TAG, "loaded ${items.size} items (built=$builtAt)")
        } catch (t: Throwable) {
            loadError = "load failed: ${t.message}"
            Log.e(TAG, "load failed", t)
        }
    }

    private fun addCameras(arr: JSONArray?, out: MutableList<EnforcementItem>) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            val kind = EnforcementKind.from(o.optString("kind")) ?: continue
            out.add(
                EnforcementItem(
                    id = o.getInt("id"), kind = kind,
                    lat = o.getDouble("lat"), lon = o.getDouble("lon"),
                    limit = o.optIntOrNull("limit"), dir = o.optStringOrNull("dir"),
                    addr = o.optStringOrNull("addr"), city = o.optStringOrNull("city")
                )
            )
        }
    }

    private fun addSections(arr: JSONArray?, out: MutableList<EnforcementItem>) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            out.add(
                EnforcementItem(
                    id = 1_000_000 + o.getInt("id"), kind = EnforcementKind.INTERVAL_SPEED,
                    lat = o.getDouble("lat"), lon = o.getDouble("lon"),
                    limit = o.optIntOrNull("limit"), dir = null,
                    addr = o.optStringOrNull("name"), city = o.optStringOrNull("city")
                )
            )
        }
    }

    private fun addAccidents(arr: JSONArray?, out: MutableList<EnforcementItem>) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            if (o.isNull("lat") || o.isNull("lon")) continue
            out.add(
                EnforcementItem(
                    id = 2_000_000 + o.getInt("id"), kind = EnforcementKind.ACCIDENT_SEGMENT,
                    lat = o.getDouble("lat"), lon = o.getDouble("lon"),
                    limit = null, dir = null,
                    addr = o.optStringOrNull("name"), city = o.optStringOrNull("city")
                )
            )
        }
    }

    private fun JSONObject.optIntOrNull(key: String): Int? =
        if (isNull(key)) null else optInt(key).takeIf { has(key) }

    private fun JSONObject.optStringOrNull(key: String): String? =
        if (isNull(key)) null else optString(key).ifBlank { null }

    private fun key(latCell: Int, lonCell: Int): Long =
        (latCell.toLong() shl 32) xor (lonCell.toLong() and 0xffffffffL)

    private fun buildIndex() {
        grid.clear()
        items.forEachIndexed { i, it ->
            val k = key(floor(it.lat / cell).toInt(), floor(it.lon / cell).toInt())
            grid.getOrPut(k) { ArrayList() }.add(i)
        }
    }

    /** Items within [radius] metres of (lat,lon), nearest first. */
    fun nearby(
        lat: Double, lon: Double, radius: Double,
        kinds: Set<EnforcementKind> = EnforcementKind.entries.toSet()
    ): List<NearbyHit> {
        val latSpan = ceil(radius / (cell * metersPerDegLat)).toInt() + 1
        val lonScale = max(cos(Math.toRadians(lat)), 0.1)
        val lonSpan = ceil(radius / (cell * metersPerDegLat * lonScale)).toInt() + 1
        val latCell = floor(lat / cell).toInt()
        val lonCell = floor(lon / cell).toInt()

        val out = ArrayList<NearbyHit>()
        val seen = HashSet<Int>()
        for (dy in -latSpan..latSpan) {
            for (dx in -lonSpan..lonSpan) {
                val ids = grid[key(latCell + dy, lonCell + dx)] ?: continue
                for (i in ids) {
                    if (!seen.add(i)) continue
                    val it = items[i]
                    if (it.kind !in kinds) continue
                    val d = Geo.distance(lat, lon, it.lat, it.lon)
                    if (d <= radius) out.add(NearbyHit(it, d, Geo.bearing(lat, lon, it.lat, it.lon)))
                }
            }
        }
        out.sortBy { it.distance }
        return out
    }

    private companion object { const val TAG = "SpeedAlert" }
}

File: ios/SpeedAlert/Data/EnforcementStore.swift
import Foundation
import CoreLocation

/// Loads the bundled offline dataset and answers proximity queries.
///
/// Primary source is a **binary property list** (`app_data.plist`) because
/// Foundation's JSON parser rejected our (otherwise valid) UTF-8 JSON on
/// iOS 26. Falls back to `app_data.json`. Never crashes: records `loadError`.
final class EnforcementStore: ObservableObject {
    @Published private(set) var itemCount: Int = 0
    @Published private(set) var builtAt: String?
    @Published private(set) var loadError: String?
    @Published private(set) var resourceFound = false

    private var items: [EnforcementItem] = []
    private var grid: [Int64: [Int]] = [:]          // cellKey -> indices
    private let cell = 0.02                          // ~2.2 km cells
    private let metersPerDegLat = 111_320.0

    init(bundle: Bundle = .main) { load(bundle: bundle) }

    func load(bundle: Bundle) {
        // 1) binary plist (primary)
        if let url = bundle.url(forResource: "app_data", withExtension: "plist"),
           let data = try? Data(contentsOf: url) {
            resourceFound = true
            if decode(data, usingPlist: true, name: url.lastPathComponent) { return }
        }
        // 2) JSON (fallback)
        if let url = bundle.url(forResource: "app_data", withExtension: "json"),
           let data = try? Data(contentsOf: url) {
            resourceFound = true
            if decode(data, usingPlist: false, name: url.lastPathComponent) { return }
        }
        if !resourceFound {
            loadError = "app_data not in bundle"
            NSLog("[SpeedAlert] MISSING app_data; json=%@",
                  bundle.paths(forResourcesOfType: "json", inDirectory: nil).description)
        }
    }

    private func decode(_ data: Data, usingPlist: Bool, name: String) -> Bool {
        do {
            let raw: RawAppData
            if usingPlist {
                raw = try PropertyListDecoder().decode(RawAppData.self, from: data)
            } else {
                raw = try JSONDecoder().decode(RawAppData.self, from: data)
            }
            builtAt = raw.builtAt
            var all = raw.cameras.map { $0.toItem() }
            all.append(contentsOf: raw.sections.map { $0.toItem() })
            all.append(contentsOf: raw.accidents.compactMap { $0.toItem() })
            items = all
            buildIndex()
            itemCount = items.count
            loadError = nil
            NSLog("[SpeedAlert] loaded %d items from %@ (cameras=%d sections=%d accidents=%d)",
                  itemCount, name, raw.cameras.count, raw.sections.count, raw.accidents.count)
            return true
        } catch {
            let d = EnforcementStore.describe(error)
            loadError = "(\(name)) " + d
            NSLog("[SpeedAlert] %@ decode FAILED: %@ (count=%d)", name, d, data.count)
            return false
        }
    }

    /// Detailed, human-readable description of a decoding error (incl. codingPath).
    static func describe(_ error: Error) -> String {
        guard let de = error as? DecodingError else { return "\(error)" }
        func joined(_ path: [CodingKey]) -> String {
            return path.map { $0.stringValue }.joined(separator: ".")
        }
        switch de {
        case .dataCorrupted(let ctx):          return "dataCorrupted @[\(joined(ctx.codingPath))] :: \(ctx.debugDescription)"
        case .keyNotFound(let key, let ctx):   return "keyNotFound(\(key.stringValue)) @[\(joined(ctx.codingPath))]"
        case .typeMismatch(let type, let ctx): return "typeMismatch(\(type)) @[\(joined(ctx.codingPath))]"
        case .valueNotFound(let type, let ctx): return "valueNotFound(\(type)) @[\(joined(ctx.codingPath))]"
        @unknown default:                      return "decoding error \(de)"
        }
    }

    private func cellKey(_ latCell: Int, _ lonCell: Int) -> Int64 {
        (Int64(latCell) << 32) ^ Int64(UInt32(bitPattern: Int32(lonCell)))
    }

    private func buildIndex() {
        grid.removeAll(keepingCapacity: true)
        for (i, it) in items.enumerated() {
            let key = cellKey(Int(floor(it.lat / cell)), Int(floor(it.lon / cell)))
            grid[key, default: []].append(i)
        }
    }

    /// Items within `radius` metres of `coord`, nearest first.
    func nearby(_ coord: CLLocationCoordinate2D,
                radius: CLLocationDistance,
                kinds: Set<EnforcementKind> = Set(EnforcementKind.allCases)) -> [NearbyHit] {
        let origin = CLLocation(latitude: coord.latitude, longitude: coord.longitude)
        let latSpan = Int(ceil(radius / (cell * metersPerDegLat))) + 1
        let lonScale = max(cos(coord.latitude * .pi / 180), 0.1)
        let lonSpan = Int(ceil(radius / (cell * metersPerDegLat * lonScale))) + 1
        let latCell = Int(floor(coord.latitude / cell))
        let lonCell = Int(floor(coord.longitude / cell))

        var hits: [NearbyHit] = []
        var seen = Set<Int>()
        for dy in -latSpan...latSpan {
            for dx in -lonSpan...lonSpan {
                guard let idxs = grid[cellKey(latCell + dy, lonCell + dx)] else { continue }
                for i in idxs where seen.insert(i).inserted {
                    let it = items[i]
                    guard kinds.contains(it.kind) else { continue }
                    let d = origin.distance(from: CLLocation(latitude: it.lat, longitude: it.lon))
                    guard d <= radius else { continue }
                    hits.append(NearbyHit(item: it, distance: d,
                                          bearing: EnforcementStore.bearing(from: coord, to: it.coordinate)))
                }
            }
        }
        hits.sort { $0.distance < $1.distance }
        return hits
    }

    /// Initial bearing from one coordinate to another (degrees, 0 = north).
    static func bearing(from a: CLLocationCoordinate2D, to b: CLLocationCoordinate2D) -> Double {
        let p1 = a.latitude * .pi / 180, p2 = b.latitude * .pi / 180
        let dl = (b.longitude - a.longitude) * .pi / 180
        let y = sin(dl) * cos(p2)
        let x = cos(p1) * sin(p2) - sin(p1) * cos(p2) * cos(dl)
        return (atan2(y, x) * 180 / .pi + 360).truncatingRemainder(dividingBy: 360)
    }
}
